python_modules/dagster/dagster/_core/definitions/metadata/source_code.py

```diff
@@ -0,0 +1,145 @@
+import inspect
+import os
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    Callable,
+    Mapping,
+    Optional,
+    Sequence,
+    Union,
+)
+
+import pydantic
+
+import dagster._check as check
+from dagster._annotations import experimental
+from dagster._model import DagsterModel
+from dagster._serdes import whitelist_for_serdes
+
+from .metadata_set import (
+    NamespacedMetadataSet as NamespacedMetadataSet,
+    TableMetadataSet as TableMetadataSet,
+)
+from .metadata_value import MetadataValue
+
+if TYPE_CHECKING:
+    from dagster._core.definitions.assets import AssetsDefinition, SourceAsset
+    from dagster._core.definitions.cacheable_assets import CacheableAssetsDefinition
+
+DEFAULT_SOURCE_FILE_KEY = "asset_definition"
+
+
+@experimental
+@whitelist_for_serdes
+class LocalFileCodeReference(DagsterModel):
+    """Represents a local file source location."""
+
+    file_path: str
+    line_number: int
+
+
+@experimental
+@whitelist_for_serdes
+class CodeReferencesMetadataValue(DagsterModel, MetadataValue["CodeReferencesMetadataValue"]):
+    """Metadata value type which represents source locations (locally or otherwise)
+    of the asset in question. For example, the file path and line number where the
+    asset is defined.
+
+    Attributes:
+        sources (Dict[str, LocalFileCodeReference]):
+            A labeled dictionary of sources. The main source file should be keyed with
+            the `DEFAULT_SOURCE_FILE_KEY` constant.
+    """
+
+    code_references: Mapping[str, LocalFileCodeReference]
```

Comment:
You should describe what keys represent and have an example.

Also this `DEFAULT_SOURCE_FILE_KEY` thing is very fishy. Are we expecting users to import that in the common case?